### PR TITLE
BUG-1278: Introduce boolean nostream parameter to StatsCmd.

### PR DESCRIPTION
**Pull request by alexey-atiskov**
Fix for https://github.com/docker-java/docker-java/issues/1278

---
This change is [Reviewable](https://reviewable.io/reviews/docker-java/docker-java/1287)

### CODE REVIEW

File: docker-java-core/src/main/java/com/github/dockerjava/core/exec/StatsCmdExec.java [hunk 1]
```diff
@@ -23,6 +25,10 @@ protected Void execute0(StatsCmd command, ResultCallback<Statistics> resultCallb
         WebTarget webTarget = getBaseResource().path("/containers/{id}/stats").resolveTemplate("id",
                 command.getContainerId());
 
+        if (isTrue(command.getNoStream())) {
+            webTarget = webTarget.queryParam("stream", "0");
```
Comment (KostyaSha) [inline docker-java-core/src/main/java/com/github/dockerjava/core/exec/StatsCmdExec.java]: what 0 mean here?

Comment (alexey-atiskov) [reply docker-java-core/src/main/java/com/github/dockerjava/core/exec/StatsCmdExec.java]: This is only equivalent to curls, such as 
curl https://172.17.42.1:2375/containers/example/stats?stream=0
curl https://172.17.42.1:2375/containers/example/stats?stream=false

Comment (alexey-atiskov) [reply docker-java-core/src/main/java/com/github/dockerjava/core/exec/StatsCmdExec.java]: Also I saw parsing in https://github.com/moby/moby/blob/master/client/container_stats.go

File: docker-java-api/src/main/java/com/github/dockerjava/api/command/StatsCmd.java [hunk 2]
```diff
@@ -16,6 +16,11 @@
 
     StatsCmd withContainerId(@Nonnull String containerId);
 
+    @CheckForNull
+    Boolean getNoStream();
+
+    StatsCmd withNoStream(@Nonnull Boolean noStream);
```
Comment (bsideup) [inline docker-java-api/src/main/java/com/github/dockerjava/api/command/StatsCmd.java]:
```suggestion
    StatsCmd withNoStream(boolean noStream);
```

File: docker-java-core/src/main/java/com/github/dockerjava/core/exec/StatsCmdExec.java [hunk 3]
```diff
@@ -23,6 +25,10 @@ protected Void execute0(StatsCmd command, ResultCallback<Statistics> resultCallb
         WebTarget webTarget = getBaseResource().path("/containers/{id}/stats").resolveTemplate("id",
                 command.getContainerId());
 
+        if (isTrue(command.getNoStream())) {
```
Comment (bsideup) [inline docker-java-core/src/main/java/com/github/dockerjava/core/exec/StatsCmdExec.java]: I believe we don't need `BooleanUtils` here and can simply use `Boolean.TRUE.equals(command.getNoStream())` (null safe)

File: docker-java/src/test/java/com/github/dockerjava/cmd/StatsCmdIT.java [hunk 4]
```diff
@@ -56,6 +57,40 @@ public void testStatsStreaming() throws InterruptedException, IOException {
 
     }
 
+    @Test
+    public void testStatsNoStreaming() throws InterruptedException, IOException {
+        TimeUnit.SECONDS.sleep(1);
```
Comment (bsideup) [inline docker-java/src/test/java/com/github/dockerjava/cmd/StatsCmdIT.java]: why is it here?

File: docker-java/src/test/java/com/github/dockerjava/cmd/StatsCmdIT.java [hunk 5]
```diff
@@ -56,6 +57,40 @@ public void testStatsStreaming() throws InterruptedException, IOException {
...
+
+        StatsCallbackTest statsCallback = dockerRule.getClient().statsCmd(container.getId()).withNoStream(true).exec(
+                new StatsCallbackTest(countDownLatch));
+
+        countDownLatch.await(5, TimeUnit.SECONDS);
+        Boolean gotStats = statsCallback.gotStats();
+
+        LOG.info("Stop stats collection");
+
+        statsCallback.close();
```
Comment (bsideup) [inline docker-java/src/test/java/com/github/dockerjava/cmd/StatsCmdIT.java]: please use `try-with-resources` to avoid leaks

File: docker-java/src/test/java/com/github/dockerjava/cmd/StatsCmdIT.java (hunk 4 above)
Comment (alexey-atiskov) [reply docker-java/src/test/java/com/github/dockerjava/cmd/StatsCmdIT.java]: Same as in testStatsStreaming() method. Maybe already not needed in both places.

File: docker-java/src/test/java/com/github/dockerjava/cmd/StatsCmdIT.java (hunk 5 above)
Comment (alexey-atiskov) [reply docker-java/src/test/java/com/github/dockerjava/cmd/StatsCmdIT.java]: Seems like all tests should be reviewed for such cases. I have tried to make them "similar"...

File: docker-java/src/test/java/com/github/dockerjava/cmd/StatsCmdIT.java (hunk 4 above)
Comment (bsideup) [reply docker-java/src/test/java/com/github/dockerjava/cmd/StatsCmdIT.java]: yeah, I think so. Especially at the beginning of the test where everything should already be prepared :)

File: docker-java-api/src/main/java/com/github/dockerjava/api/command/StatsCmd.java [hunk 6]
```diff
@@ -16,6 +16,11 @@
 
     StatsCmd withContainerId(@Nonnull String containerId);
 
+    @CheckForNull
+    Boolean getNoStream();
```
Comment (bsideup) [inline docker-java-api/src/main/java/com/github/dockerjava/api/command/StatsCmd.java]: please follow the naming of other boolean properties as seen in other commands (e.g. `AttachContainerCmd`)

File: docker-java-core/src/main/java/com/github/dockerjava/core/exec/StatsCmdExec.java (hunk 1 above)
Comment (bsideup) [reply docker-java-core/src/main/java/com/github/dockerjava/core/exec/StatsCmdExec.java]: @alexey-atiskov please do not resolve conversations that were started by somebody else. It is up to the reviewer to decide whether something is resolved or not

File: docker-java/src/test/java/com/github/dockerjava/cmd/StatsCmdIT.java [hunk 7]
```diff
@@ -37,23 +36,52 @@ public void testStatsStreaming() throws InterruptedException, IOException {
...
+
+        LOG.info("Completed test");
+        assertTrue("Expected true", gotStats);
+    }
+
+    @Test
+    public void testStatsNoStreaming() throws InterruptedException, IOException {
+        CountDownLatch countDownLatch = new CountDownLatch(NUM_STATS);
+
+        String containerName = "generated_" + new SecureRandom().nextInt();
```
Comment (bsideup) [inline docker-java/src/test/java/com/github/dockerjava/cmd/StatsCmdIT.java]: Unless you need to assert the name, you don't need to set the name and it will be auto generated anyways.
Other tests will be cleaned up soon as well.

File: docker-java/src/test/java/com/github/dockerjava/cmd/StatsCmdIT.java [hunk 8]
```diff
@@ -37,23 +36,52 @@ public void testStatsStreaming() throws InterruptedException, IOException {
...
+
+        CreateContainerResponse container = dockerRule.getClient().createContainerCmd("busybox").withCmd("top")
+                .withName(containerName).exec();
+        LOG.info("Created container {}", container.toString());
+        assertThat(container.getId(), not(isEmptyString()));
```
Comment (bsideup) [inline docker-java/src/test/java/com/github/dockerjava/cmd/StatsCmdIT.java]: this is tested in `CreateContainerCmdIT`, no need to log/assert it

File: docker-java/src/test/java/com/github/dockerjava/cmd/StatsCmdIT.java [hunk 9]
```diff
@@ -37,23 +36,52 @@ public void testStatsStreaming() throws InterruptedException, IOException {
...
-        statsCallback.close();
+            LOG.info("Stop stats collection");
+        }
 
         LOG.info("Stopping container");
         dockerRule.getClient().stopContainerCmd(container.getId()).exec();
         dockerRule.getClient().removeContainerCmd(container.getId()).exec();
 
         LOG.info("Completed test");
         assertTrue("Expected true", gotStats);
```
Comment (bsideup) [inline docker-java/src/test/java/com/github/dockerjava/cmd/StatsCmdIT.java]: already tested in `testStatsStreaming`

File: docker-java/src/test/java/com/github/dockerjava/cmd/StatsCmdIT.java [hunk 10]
```diff
@@ -37,23 +36,52 @@ public void testStatsStreaming() throws InterruptedException, IOException {
...
+                .withName(containerName).exec();
+        LOG.info("Created container {}", container.toString());
+        assertThat(container.getId(), not(isEmptyString()));
 
-        countDownLatch.await(3, TimeUnit.SECONDS);
-        Boolean gotStats = statsCallback.gotStats();
+        dockerRule.getClient().startContainerCmd(container.getId()).exec();
 
-        LOG.info("Stop stats collection");
+        Boolean gotStats;
```
Comment (bsideup) [inline docker-java/src/test/java/com/github/dockerjava/cmd/StatsCmdIT.java]:
```suggestion
        boolean gotStats = false;
```

File: .idea/encodings.xml [hunk 11]
```diff
@@ -7,6 +7,7 @@
     <file url="file://$PROJECT_DIR$/docker-java-core" charset="UTF-8" />
     <file url="file://$PROJECT_DIR$/docker-java-transport-jersey" charset="UTF-8" />
     <file url="file://$PROJECT_DIR$/docker-java-transport-netty" charset="UTF-8" />
+    <file url="file://$PROJECT_DIR$/docker-java-transport-okhttp" charset="UTF-8" />
```
Comment (bsideup) [inline .idea/encodings.xml]: could you please revert this change? This file is going to be ignored and the change is unrelated to the PR anyways

**Comment by bsideup** [comment]
@alexey-atiskov please merge with master, there are conflicts

File: .idea/encodings.xml (hunk 11 above)
Comment (bsideup) [reply .idea/encodings.xml]: Also see: https://github.com/docker-java/docker-java/pull/1312

File: docker-java-api/src/main/java/com/github/dockerjava/api/command/StatsCmd.java (hunk 6 above)
Comment (alexey-atiskov) [reply docker-java-api/src/main/java/com/github/dockerjava/api/command/StatsCmd.java]: Renamed to hasNoStream

File: docker-java/src/test/java/com/github/dockerjava/cmd/StatsCmdIT.java (hunk 7 above)
Comment (alexey-atiskov) [reply docker-java/src/test/java/com/github/dockerjava/cmd/StatsCmdIT.java]: Removed setting name explicitly.

File: docker-java/src/test/java/com/github/dockerjava/cmd/StatsCmdIT.java (hunk 8 above)
Comment (alexey-atiskov) [reply docker-java/src/test/java/com/github/dockerjava/cmd/StatsCmdIT.java]: Removed assertions and logs.

File: docker-java/src/test/java/com/github/dockerjava/cmd/StatsCmdIT.java (hunk 9 above)
Comment (alexey-atiskov) [reply docker-java/src/test/java/com/github/dockerjava/cmd/StatsCmdIT.java]: Removed assertion.

File: docker-java/src/test/java/com/github/dockerjava/cmd/StatsCmdIT.java (hunk 10 above)
Comment (alexey-atiskov) [reply docker-java/src/test/java/com/github/dockerjava/cmd/StatsCmdIT.java]: Done. (=false is only for readability)

File: .idea/encodings.xml (hunk 11 above)
Comment (alexey-atiskov) [reply .idea/encodings.xml]: Agree. My IDEA updates this file silently =(

File: docker-java-core/src/main/java/com/github/dockerjava/core/exec/StatsCmdExec.java (hunk 1 above)
Comment (alexey-atiskov) [reply docker-java-core/src/main/java/com/github/dockerjava/core/exec/StatsCmdExec.java]: Oh, shame on me, I noticed this comment tool late(

Comment (bsideup) [reply docker-java-core/src/main/java/com/github/dockerjava/core/exec/StatsCmdExec.java]: no problem :)
